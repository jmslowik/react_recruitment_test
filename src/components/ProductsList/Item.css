.row {
  padding: 10px;
  border-bottom: 1px solid silver;
  list-style: none;
  display: flex;
}

.after {
  margin-left: auto;
}

.price {
  margin-left: auto;
}

.item {
  width: 60%;
  display: flex;
}