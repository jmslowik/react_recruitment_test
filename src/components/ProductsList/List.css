.container{
    width: 1000px;
    margin: 100px auto 0 auto;
}

.list {
    margin-bottom: 0;
    border-bottom-style: solid;
}

.list .row:last-child {
    border-bottom-style: none;
}

.sum {
    font-weight: bold;
}