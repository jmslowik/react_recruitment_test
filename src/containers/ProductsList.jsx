import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getAllAction } from '@/store/products/saga';
import {
  setAndVerifyProductQuantityAction,
  setAndVerifyProductQuantityDebounceAction,
} from '@/store/basket/saga';
import { productsListSelector } from '@/store/products/selectors';
import { basketSelector } from '@/store/basket/selectors';
import ProductsListComponent from '@/components/ProductsList';

const ProductsList = () => {
  const dispatch = useDispatch();

  const products = useSelector(productsListSelector);
  const basket = useSelector(basketSelector);

  const onChange = ({ pid, newValue, min }) => {
    dispatch(setAndVerifyProductQuantityDebounceAction({ pid, quantity: newValue, min }));
  };

  useEffect(() => {
    dispatch(getAllAction());
  }, [dispatch]);

  useEffect(() => {
    products.forEach(({ pid, min }) => {
      dispatch(setAndVerifyProductQuantityAction({ pid, quantity: min, min }));
    });
  }, [products, dispatch]);

  return (
    <ProductsListComponent
      list={products}
      quantities={basket}
      onChange={onChange}
    />
  );
};

export default ProductsList;
