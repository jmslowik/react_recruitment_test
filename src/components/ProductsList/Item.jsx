import React from 'react';
import './Item.css';

// TODO locale and currency as prop

const Item = ({
  name, price = 0, after,
}) => (
  <li className="row">
    <div className="item">
      {`${name}:`}
      <span className="price">
        {price.toLocaleString('pl-PL', { style: 'currency', currency: 'PLN' })}
      </span>

    </div>
    <div className="after">{after}</div>
  </li>
);

export default Item;
