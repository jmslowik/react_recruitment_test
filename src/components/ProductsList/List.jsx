import React from 'react';
import './List.css';

// TODO name as prop

const List = ({ children, sum }) => (
  <div className="container">
    <h3>Lista produkt√≥w</h3>
    <ul className="list">
      { children }
    </ul>
    <ul className="sum">
      { sum }
    </ul>
  </div>
);

export default List;
