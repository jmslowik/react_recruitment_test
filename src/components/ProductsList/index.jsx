import React from 'react';
import List from '@/components/ProductsList/List';
import Item from '@/components/ProductsList/Item';
import QuantityController from '@/components/QuantityController';
import { parsePrice, getSum, getTotalQuantity } from '@/utils';

const ProductsList = ({ list, quantities, onChange }) => {
  const totalPrice = getSum(list, quantities);
  const totalQuantity = getTotalQuantity(quantities);

  const Sum = () => (
    <Item
      name="Suma"
      price={totalPrice}
      after={<div>{`Obecnie masz ${totalQuantity} przedmiotów w koszyku`}</div>}
    />
  );

  return (
    <List name="Lista produktów" sum={<Sum />}>
      { list && list.map(({
        pid, name, price, min, max, isBlocked,
      }) => (
        quantities[pid] && (
        <Item
          key={pid}
          name={name}
          price={parsePrice(price)}
          after={(
            <QuantityController
              quantity={(quantities[pid] && quantities[pid].quantity) || min}
              min={min}
              max={max}
              isBlocked={isBlocked}
              onChange={(newValue) => onChange({ pid, newValue, min })}
            />
          )}
        />
        )
      ))}
    </List>
  );
};

export default ProductsList;
