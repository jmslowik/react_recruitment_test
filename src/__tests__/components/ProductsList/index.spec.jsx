import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import ProductsList from '@/components/ProductsList/index';

describe('ProductsList', () => {
  it('should render product list', () => {
    const wrapper = mount(<ProductsList />);
    expect(wrapper.contains(<h3>Lista produkt√≥w</h3>)).toBeTruthy();
  });

  it('should render correctly', () => {
    const tree = renderer.create(<ProductsList />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
