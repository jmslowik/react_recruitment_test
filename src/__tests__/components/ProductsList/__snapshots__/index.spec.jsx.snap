// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList should render correctly 1`] = `
<div
  className="container"
>
  <h3>
    Lista produktów
  </h3>
  <ul
    className="list"
  />
  <ul
    className="sum"
  >
    <li
      className="row"
    >
      <div
        className="item"
      >
        Suma:
        <span
          className="price"
        >
          0,00 zł
        </span>
      </div>
      <div
        className="after"
      >
        <div>
          Obecnie masz 0 przedmiotów w koszyku
        </div>
      </div>
    </li>
  </ul>
</div>
`;
